### PR TITLE
Receive from one named session

### PR DESCRIPTION
**Pull request by hemanttanwar**
receive from one session id

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/pom.xml [hunk 1]
```diff
@@ -42,7 +42,7 @@
     <dependency>
       <groupId>com.azure</groupId>
       <artifactId>azure-core-amqp</artifactId>
-      <version>1.0.2</version> <!-- {x-version-update;com.azure:azure-core-amqp;dependency} -->
+      <version>1.1.0-beta.1</version> <!-- {x-version-update;unreleased_com.azure:azure-core-amqp;dependency} -->
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/pom.xml:45]: These changes should not be merged until we release EH.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java [hunk 2]
```diff
@@ -356,7 +355,7 @@ protected ReactorReceiver createConsumer(String entityPath, Receiver receiver,
      * NOTE: Ensure this is invoked using the reactor dispatcher because proton-j is not thread-safe.
      */
     private LinkSubscription<AmqpReceiveLink> getSubscription(String linkName, String entityPath,
-        Map<Symbol, UnknownDescribedType> sourceFilters, Map<Symbol, Object> receiverProperties,
+        Map<Symbol, Object> sourceFilters, Map<Symbol, Object> receiverProperties,
```
Comment (conniey) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java:358]: Why the change to object?

Comment (conniey) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java:358]: You can translate an object into that type.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java [hunk 3]
```diff
@@ -105,7 +106,7 @@
      * @param onClientClose Operation to run when the client completes.
      */
     ServiceBusReceiverAsyncClient(String fullyQualifiedNamespace, String entityPath, MessagingEntityType entityType,
-        boolean isSessionEnabled, ReceiverOptions receiverOptions,
+        boolean isSessionEnabled, String sessionId, ReceiverOptions receiverOptions,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java:109]: I'd dump it into receiver options. And probably pull in is session enabled, too

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 4]
```diff
@@ -44,5 +44,5 @@
      * @return A new or existing receive link that is connected to the given {@code entityPath}.
      */
     Mono<AmqpReceiveLink> createReceiveLink(String linkName, String entityPath, ReceiveMode receiveMode,
-        boolean isSession, String transferEntityPath, MessagingEntityType entityType);
+        boolean isSession, String sessionId, String transferEntityPath, MessagingEntityType entityType);
 }
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:48]: I know we had unused sessionenabled. But add another method rather than everything in this method

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ReceiverOptions.java [hunk 5]
```diff
@@ -27,10 +29,43 @@ ReceiveMode getReceiveMode() {
     }
 
     /**
-     * Gets the prefetch count of the receiver.
+     * Update enableSession flag.
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ReceiverOptions.java:32]: This doesn't need to be a mutable object. That's why the other fields are read-only.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 6]
```diff
@@ -593,8 +606,8 @@ public ServiceBusReceiverAsyncClient buildAsyncClient() {
             final ReceiverOptions receiverOptions = new ReceiverOptions(receiveMode, prefetchCount);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:606]: The options don't seem to be propagated here.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java [hunk 7]
```diff
@@ -121,7 +121,8 @@
         this.receiveMode = receiverOptions.getReceiveMode();
 
         this.entityType = entityType;
-        this.isSessionEnabled = isSessionEnabled;
+        this.isSessionEnabled = receiverOptions.isSessionEnabled();
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClient.java:124]: You don't need to create dedicated fields now. A reference to the receiverOptions is held by the class.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 8]
```diff
@@ -45,4 +45,23 @@
      */
     Mono<AmqpReceiveLink> createReceiveLink(String linkName, String entityPath, ReceiveMode receiveMode,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:46]: Since this is a normal link creation, we don't need isSession.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 9]
```diff
@@ -45,4 +45,23 @@
...
+     *     link.
+     * @param receiveMode {@link ReceiveMode} to use when creating the link.
+     * @param isSession to use when creating the link.
+     * @param sessionId to use when creating the link.
+     * @param transferEntityPath to use when creating the link.
+     * @param entityType {@link MessagingEntityType} to use when creating the link.
+     *
+     * @return A new or existing receive link that is connected to the given {@code entityPath}.
+     */
+    Mono<AmqpReceiveLink> createReceiveLink(String linkName, String entityPath, ReceiveMode receiveMode,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:65]: This would be an overload where you don't need "isSession" because we'd have a sessionId. And the parameter would be at the end.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java [hunk 10]
```diff
@@ -66,14 +66,14 @@
 
     @Override
     public Mono<AmqpReceiveLink> createConsumer(String linkName, String entityPath, MessagingEntityType entityType,
-        Duration timeout, AmqpRetryPolicy retry, ReceiveMode receiveMode, boolean isSession) {
+        Duration timeout, AmqpRetryPolicy retry, ReceiveMode receiveMode, boolean isSession, String sessionId) {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java:69]: I'd be inclined to create another method createConsumer and remove the isSession. One createconsumer for normal receives, one for sessions, and another one to support transactions in the future.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 11]
```diff
@@ -590,11 +603,12 @@ public ServiceBusReceiverAsyncClient buildAsyncClient() {
 
             final MessageLockContainer messageLockContainer = new MessageLockContainer();
             final ServiceBusConnectionProcessor connectionProcessor = getOrCreateConnectionProcessor(messageSerializer);
-            final ReceiverOptions receiverOptions = new ReceiverOptions(receiveMode, prefetchCount);
+            final ReceiverOptions receiverOptions = new ReceiverOptions(receiveMode, prefetchCount,
+                false, sessionId);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:607]: Why is it always false? If later on, you're inferring it from the presence of a `sessionId`, the "enableSession" isn't required.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 12]
```diff
@@ -44,5 +44,23 @@
      * @return A new or existing receive link that is connected to the given {@code entityPath}.
      */
     Mono<AmqpReceiveLink> createReceiveLink(String linkName, String entityPath, ReceiveMode receiveMode,
-        boolean isSession, String transferEntityPath, MessagingEntityType entityType);
+        String transferEntityPath, MessagingEntityType entityType);
+
+    /**
+     * Creates or gets an existing receive link. The same link is returned if there is an existing receive link with the
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:50]: Doc update: Creates a receive link for sessions. 
Since this is similar to the method above, I had mentioned to add the "SessionId" to the end of the parameter list as an overload. This would be easier to consume.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java [hunk 13]
```diff
@@ -159,14 +159,34 @@ public ServiceBusReactorAmqpConnection(String connectionId, ConnectionOptions co
      */
     @Override
     public Mono<AmqpReceiveLink> createReceiveLink(String linkName, String entityPath, ReceiveMode receiveMode,
-        boolean isSession, String transferEntityPath, MessagingEntityType entityType) {
+        String transferEntityPath, MessagingEntityType entityType) {
+        return createReceiveLink(linkName, entityPath, receiveMode, null,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java:163]: Since these are performing two different operations, and you have created a separate method in ReactorSession for creating a consumer with a sessionId. This should call that.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java [hunk 14]
```diff
@@ -24,11 +24,13 @@
 
 class ServiceBusReceiverAsyncClientIntegrationTest extends IntegrationTestBase {
     private static final String CONTENTS = "Test-contents";
+    private static final String SESSION_ID = "seattle-id1";
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java:27]: This wouldn't work because we need a session enabled queue. You'd have to update the ARM template to create one.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java [hunk 15]
```diff
@@ -233,8 +232,8 @@ void receivesAndAutoCompletes() {
         // Arrange
         final ReceiverOptions options = new ReceiverOptions(ReceiveMode.PEEK_LOCK, PREFETCH);
         final ServiceBusReceiverAsyncClient consumer2 = new ServiceBusReceiverAsyncClient(
-            NAMESPACE, ENTITY_PATH, MessagingEntityType.QUEUE, false, options, connectionProcessor,
-            tracerProvider, messageSerializer, messageContainer, onClientClose);
+            NAMESPACE, ENTITY_PATH, MessagingEntityType.QUEUE, options, connectionProcessor,  tracerProvider,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientTest.java:235]: Reformat the code. There are two spaces here.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ReceiverOptions.java [hunk 16]
```diff
@@ -11,10 +11,16 @@
 class ReceiverOptions {
     private final ReceiveMode receiveMode;
     private final int prefetchCount;
+    private final String sessionId;
 
     ReceiverOptions(ReceiveMode receiveMode, int prefetchCount) {
+        this(receiveMode, prefetchCount, null);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ReceiverOptions.java:17]: Why do we need another constructor? This is only used in SBCB to pass to the receiver.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ReceiverOptions.java [hunk 17]
```diff
@@ -27,10 +33,14 @@ ReceiveMode getReceiveMode() {
     }
 
     /**
-     * Gets the prefetch count of the receiver.
+     * Gets the sessionId for the message.
      *
-     * @return The prefetch count of the receiver.
+     * @return the sessionId for the message.
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ReceiverOptions.java:38]: "Id of the session to receive messages from."
